### PR TITLE
blockchain/vm: refactor gas table (from geth #19735)

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes

- This PR is derived from https://github.com/ethereum/go-ethereum/pull/19735.
- It refactors `blockchain/vm/gas_table.go` and add constantGas value to some opcodes.

Added `blockchain/vm/gas_table.go` difference between klaytn and geth code.
- gasSStore: deleted eip1283 part (petersburg)
- gasCall: deleted eip158 part (tangerine whistle)
- gasSuicide: deleted eip150, eip158 part (tangerine whistle)
- gasExp: integrate gasExpFrontier and gasExpEIP158 to EIP158 (tangerine whistle)

Since klaytn is forked after constantinople, all previous hard fork versions such as tangerine whistle are default. Petersburg came out after constantinople, but klaytn included it before releasing it to public, so this is also default.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by ehnuje — DISMISSED** [review]
LGTM

**Comment by yoomee1313** [comment]
I ran baobab sync test for sure :). This PR made it past all blocks of baobab testnet. Please take a look @aidan-kwon @KimKyungup @winnie-byun.

File: blockchain/vm/gas.go
Comment (aidan-kwon) [inline blockchain/vm/gas.go:41]:
```suggestion
```

File: blockchain/vm/gas.go
Comment (aidan-kwon) [inline blockchain/vm/gas.go:51]: The same condition is being checked twice.

File: blockchain/vm/gas.go
Comment (yoomee1313) [reply blockchain/vm/gas.go:41]: @aidan-kwon There were comments to record that the code related to the hard fork version before Constantinople(EIP150, EIP158) was deleted, so I left it for consistency. What do you think?

File: blockchain/vm/gas.go
Comment (yoomee1313) [reply blockchain/vm/gas.go:51]: umm.. you're right. 
original code is this. https://github.com/ethereum/go-ethereum/blob/6ec15610443b28eabf665199f1dc5be2b3e3f7cb/core/vm/gas.go#L37-L53.
This code is executed when the node's hardfork version is lower than eip150. However, klaytn's node can only have constantinople version, so this code is useless. I will remove this code.

**Review by ehnuje — DISMISSED** [review]

File: blockchain/vm/gas_table.go
Comment (winnie-byun) [inline blockchain/vm/gas_table.go:221]: The original default value was 700 and now it is chnaged to 0. Is there something changed is gas value?

File: blockchain/vm/gas_table.go
Comment (winnie-byun) [inline blockchain/vm/gas_table.go:317]: It looks like these codes are for EIP150 and EIP158. Does Klaytn support them?

**Comment by KimKyungup** [comment]
@yoomee1313 Please check review comments.

File: blockchain/vm/gas.go
Comment (yoomee1313) [reply blockchain/vm/gas.go:55]: I erased this code because it's redundant code! In geth, this code is executed when `gasTable.CreateBySuicide > 0` is false. It means that the code is only executed when the node's hardfork version is lower than tangerineWhistle. However, klaytn forked the code after the tangerineWhistle, so we don't need this logic :)

File: blockchain/vm/gas_table.go
Comment (yoomee1313) [reply blockchain/vm/gas_table.go:317]: As I said in above comment, we don't need to support those versions. Also this is a renamed version of `selfSuicide`!

File: blockchain/vm/gas_table.go
Comment (winnie-byun) [reply blockchain/vm/gas_table.go:221]: As we talked offline, it looks like this code is only used for checking enough value in wallet. It is just for estimation and 700 gas was added for surplus.
The actual gas cost is calculated in `operation.execute`.

File: blockchain/vm/gas.go
Comment (winnie-byun) [reply blockchain/vm/gas.go:55]: Thank you for the explanation. As we talked offline, `Constantinople` supports EIP150. Therfore only codes inside ` gasTable.CreateBySuicide > 0` is needed.

File: blockchain/vm/gas_table.go
Comment (winnie-byun) [inline blockchain/vm/gas_table.go:310]:
```suggestion
	gas := params.SelfdestructGas
```

**Review by winnie-byun — APPROVED** [review]

**Comment by yoomee1313** [comment]
@winnie-byun @KimKyungup @aidan-kwon @ehnuje I reviewed the comments and rebased the branch so that there are no goimports problems.

**Review by ehnuje — APPROVED** [review]

**Review by winnie-byun — APPROVED** [review]

File: blockchain/vm/gas.go
Comment (KimKyungup) [reply blockchain/vm/gas.go:56]: @yoomee1313 
This erased code could be called when `gasTable.CreateBySuicide > 0` is true also.
After EIP 150, Is there no case if `!callCost.IsUint64()` is true?

**Review by aidan-kwon — APPROVED** [review]

**Comment by aidan-kwon** [comment]
@yoomee1313 Does this PR is compatible with old versions without change gas costs?

File: blockchain/vm/gas.go
Comment (yoomee1313) [reply blockchain/vm/gas.go:56]: @KimKyungup 
I'm not sure if the answer to the question is correct. Feel free to ask a question if you have any doubts. 
`gasTable.createBySuicide > 0` is same meaning as `isEip150 == true`. You can see the changed geth code at [here](https://github.com/ethereum/go-ethereum/blob/22082f9e565665f7da045b9ff708e4698f383c77/core/vm/gas.go#L37-L53).
Since eip150 is default in klaytn, the condition is always true. Also, you can see there is only one GasTable in [here](https://github.com/klaytn/klaytn/blob/b2197d84603c67c0c99ce91ab821082d0050a608/params/gas_table.go#L44-L59).
In conclusion, the same code is executed twice every time(line 51 and line 55), so I removed the outside code of if loop.

**Comment by yoomee1313** [comment]
@aidan-kwon yes, this PR is a refactoring PR, and gas costs are same as before. I did a baobab sync test and checked that gasCall return the same gas cost as before!

**Review by KimKyungup — APPROVED** [review]

**Comment by ehnuje** [comment]
@yoomee1313 Are you waiting for review from @jeongkyun-oh ?

**Review by jeongkyun-oh — APPROVED** [review]

**Pull request merged**
